# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/FrameLayerBuilder.h
# Commit: e1edc78e9b42
# Full Hash: e1edc78e9b428352976d423c5cc5315b238f3583
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-05-27 09:54:41
# Regressor Bug: 1547624
# File Overlap Count: 2
# Description:
#   Bug Bug 1547624 - Start inactive Layer building during ProcessDisplayItems so that any contained hit-test info gets propagated to the outer Layer before we add the next item. r=miko
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D32261
# ==============================================================================

diff -r e6aef2179a23 -r e1edc78e9b42 layout/painting/FrameLayerBuilder.h
--- a/layout/painting/FrameLayerBuilder.h	Mon May 27 04:27:18 2019 +0000
+++ b/layout/painting/FrameLayerBuilder.h	Mon May 27 04:28:37 2019 +0000
@@ -40,6 +40,7 @@
 class BasicLayerManager;
 class PaintedLayer;
 class ImageLayer;
+struct LayerProperties;
 }  // namespace layers
 
 class FrameLayerBuilder;
@@ -89,6 +90,9 @@
   void SetItem(nsPaintedDisplayItem* aItem) { mItem = aItem; }
   nsPaintedDisplayItem* GetItem() const { return mItem; }
   nsIFrame* FirstFrame() const { return mFrameList[0]; }
+  layers::BasicLayerManager* InactiveManager() const {
+    return mInactiveManager;
+  }
 
   bool HasMergedFrames() const { return mFrameList.Length() > 1; }
 
@@ -223,13 +227,22 @@
   bool mIsInfinite;
 };
 
+struct InactiveLayerData {
+  RefPtr<layers::BasicLayerManager> mLayerManager;
+  FrameLayerBuilder* mLayerBuilder;
+  RefPtr<layers::Layer> mLayer;
+  UniquePtr<layers::LayerProperties> mProps;
+
+  ~InactiveLayerData();
+};
+
 struct AssignedDisplayItem {
   AssignedDisplayItem(nsPaintedDisplayItem* aItem, LayerState aLayerState,
                       DisplayItemData* aData, const nsRect& aContentRect,
                       DisplayItemEntryType aType, const bool aHasOpacity,
                       const RefPtr<TransformClipNode>& aTransform,
                       const bool aIsMerged);
-  ~AssignedDisplayItem();
+  AssignedDisplayItem(AssignedDisplayItem&& aRhs) = default;
 
   bool HasOpacity() const { return mHasOpacity; }
 
@@ -243,7 +256,7 @@
    * layer, then this stores the layer manager being
    * used for the inactive transaction.
    */
-  RefPtr<layers::LayerManager> mInactiveLayerManager;
+  UniquePtr<InactiveLayerData> mInactiveLayerData;
   RefPtr<TransformClipNode> mTransform;
 
   nsRect mContentRect;
@@ -543,7 +556,7 @@
    */
   void AddPaintedDisplayItem(PaintedLayerData* aLayerData,
                              AssignedDisplayItem& aAssignedDisplayItem,
-                             ContainerState& aContainerState, Layer* aLayer);
+                             Layer* aLayer);
 
   /**
    * Calls GetOldLayerForFrame on the underlying frame of the display item,
